reactivesocket-cpp/src/Frame.cpp

```diff
@@ -230,46 +300,63 @@ std::ostream& operator<<(std::ostream& os, const Frame_REQUEST_N& frame) {
 
 /// @{
 Payload Frame_CANCEL::serializeOut() {
-  auto buf = FrameBufferAllocator::allocate(FrameHeader::kSize);
-  folly::io::Appender app(buf.get(), /* do not grow */ 0);
-  header_.serializeInto(app);
-  return buf;
+  auto queue = FrameBufferAllocator::allocateQueue();
+  const auto metadataPresent = header_.flags_ & FrameFlags_METADATA;
+  const auto bufSize = FrameHeader::kSize +
+    (metadataPresent ? sizeof(uint32_t) : 0);
+  auto buf = FrameBufferAllocator::allocate(bufSize);
+  queue->append(std::move(buf));
+  folly::io::QueueAppender appender(queue.get(), /* do not grow */ 0);
+  header_.serializeInto(appender);
+  if (metadataPresent && metadata_.metadataPayload_) {
+    metadata_.serializeInto(appender);
+  }
+  return queue->move();
 }
 
 bool Frame_CANCEL::deserializeFrom(Payload in) {
   folly::io::Cursor cur(in.get());
   if (!header_.deserializeFrom(cur)) {
     return false;
   }
-  // TODO support metadata
-  assert((header_.flags_ & FrameFlags_METADATA) == 0);
+  if (!FrameMetadata::deserializeFrom(cur, header_.flags_, metadata_)) {
+    return false;
+  }
   return true;
 }
 
 std::ostream& operator<<(std::ostream& os, const Frame_CANCEL& frame) {
-  return os << frame.header_;
+  return os << frame.header_ << ", " << frame.metadata_;
 }
 /// @}
 
 /// @{
 Payload Frame_RESPONSE::serializeOut() {
-  auto buf = FrameBufferAllocator::allocate(FrameHeader::kSize);
-
-  folly::io::Appender app(buf.get(), /* do not grow */ 0);
-  header_.serializeInto(app);
+  auto queue = FrameBufferAllocator::allocateQueue();
+  const auto metadataPresent = header_.flags_ & FrameFlags_METADATA;
+  const auto bufSize = FrameHeader::kSize +
+    (metadataPresent ? sizeof(uint32_t) : 0);
+  auto buf = FrameBufferAllocator::allocate(bufSize);
+  queue->append(std::move(buf));
+  folly::io::QueueAppender appender(queue.get(), /* do not grow */ 0);
+  header_.serializeInto(appender);
+  if (metadataPresent && metadata_.metadataPayload_) {
```

Comment:
The "&& metadata_.metadataPayload_" part is only a precaution; the invariant is that we never have metadataPresent != metadata_.metadataPayload_ == nullptr. I used to only have metadataPresent check, but it seemed like a good idea to check both.

I will move the checks into the serialization method.